File: sample/src/main/java/com/shuhart/stepview/sample/main/MainAdapter.java
package com.shuhart.stepview.sample.main;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.shuhart.stepview.sample.R;

import org.junit.Assert;

import java.util.List;

public class MainAdapter extends RecyclerView.Adapter<MainAdapter.MainViewHolder> {

    List<Item> items;
    ItemClickListener listener;

    @NonNull
    @Override
    public MainAdapter.MainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new MainViewHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.item_main, parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull MainAdapter.MainViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull MainViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        holder.listener = null;
    }

    @Override
    public void onViewAttachedToWindow(@NonNull MainViewHolder holder) {
        super.onViewAttachedToWindow(holder);
        holder.listener = listener;
    }

    @Override
    public int getItemCount() {
        return items == null ? 0 : items.size();
    }

    static class MainViewHolder extends RecyclerView.ViewHolder {

        ItemClickListener listener;
        Item item;

        private TextView titleTextView;
        private TextView subtitleTextView;

        MainViewHolder(View itemView) {
            super(itemView);
            titleTextView = itemView.findViewById(R.id.title);
            subtitleTextView = itemView.findViewById(R.id.subtitle);
        }

        void bind(final Item item) {
            this.item = item;
            String title = null;
            String subtitle = null;
            switch (item) {
                case SIMPLE: {
                    title = titleTextView.getContext().getString(R.string.main_list_item_simple_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_simple_subtitle);
                    break;
                }
                case RECYCLER_VIEW: {
                    title = titleTextView.getContext().getString(R.string.main_list_item_recyclerview_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_recyclerview_subtitle);
                    break;
                }
                case SCROLL_VIEW: {
                    title = titleTextView.getContext().getString(R.string.main_list_item_scrollview_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_scrollview_subtitle);
                    break;
                }
                case CUSTOMISE:
                    title = titleTextView.getContext().getString(R.string.main_list_item_customise_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_customise_subtitle);
                    break;
                case RTL:
                    title = titleTextView.getContext().getString(R.string.main_list_item_rtl_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_rtl_subtitle);
                    break;
                case DELAYED_INIT:
                    title = titleTextView.getContext().getString(R.string.main_list_item_delayed_init_title);
                    subtitle = subtitleTextView.getContext().getString(R.string.main_list_item_delayed_init_subtitle);
                    break;
            }
            Assert.assertNotNull(title);
            Assert.assertNotNull(subtitle);
            titleTextView.setText(title);
            subtitleTextView.setText(subtitle);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listener != null) {
                        listener.onClick(item);
                    }
                }
            });
        }
    }

    public interface ItemClickListener {
        void onClick(Item item);
    }

    public enum Item {
        SIMPLE,
        RECYCLER_VIEW,
        SCROLL_VIEW,
        CUSTOMISE,
        RTL,
        DELAYED_INIT
    }
}

File: sample/src/main/java/com/shuhart/stepview/sample/main/MainAdapter.kt
package com.shuhart.stepview.sample.main

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.shuhart.stepview.sample.R

import junit.framework.Assert

class MainAdapter : RecyclerView.Adapter<MainAdapter.MainViewHolder>() {

    internal var items: List<Item>? = null
    internal var listener: ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainAdapter.MainViewHolder {
        return MainViewHolder(
                LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_main, parent, false)
        )
    }

    override fun onBindViewHolder(holder: MainAdapter.MainViewHolder, position: Int) {
        holder.bind(items!![position])
    }

    override fun onViewDetachedFromWindow(holder: MainViewHolder) {
        super.onViewDetachedFromWindow(holder)
        holder.listener = null
    }

    override fun onViewAttachedToWindow(holder: MainViewHolder) {
        super.onViewAttachedToWindow(holder)
        holder.listener = listener
    }

    override fun getItemCount(): Int {
        return if (items == null) 0 else items!!.size
    }

    class MainViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var listener: ItemClickListener? = null
        lateinit var item: Item

        private val titleTextView: TextView = itemView.findViewById(R.id.title)
        private val subtitleTextView: TextView = itemView.findViewById(R.id.subtitle)

        fun bind(item: Item) {
            this.item = item
            var title: String? = null
            var subtitle: String? = null
            when (item) {
                MainAdapter.Item.SIMPLE -> {
                    title = titleTextView.context.getString(R.string.main_list_item_simple_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_simple_subtitle)
                }
                MainAdapter.Item.RECYCLER_VIEW -> {
                    title = titleTextView.context.getString(R.string.main_list_item_recyclerview_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_recyclerview_subtitle)
                }
                MainAdapter.Item.SCROLL_VIEW -> {
                    title = titleTextView.context.getString(R.string.main_list_item_scrollview_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_scrollview_subtitle)
                }
                MainAdapter.Item.CUSTOMISE -> {
                    title = titleTextView.context.getString(R.string.main_list_item_customise_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_customise_subtitle)
                }
                MainAdapter.Item.RTL -> {
                    title = titleTextView.context.getString(R.string.main_list_item_rtl_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_rtl_subtitle)
                }
                MainAdapter.Item.DELAYED_INIT -> {
                    title = titleTextView.context.getString(R.string.main_list_item_delayed_init_title)
                    subtitle = subtitleTextView.context.getString(R.string.main_list_item_delayed_init_subtitle)
                }
            }
            Assert.assertNotNull(title)
            Assert.assertNotNull(subtitle)
            titleTextView.text = title
            subtitleTextView.text = subtitle
            itemView.setOnClickListener {
                if (listener != null) {
                    listener!!.onClick(item)
                }
            }
        }
    }

    interface ItemClickListener {
        fun onClick(item: Item)
    }

    enum class Item {
        SIMPLE,
        RECYCLER_VIEW,
        SCROLL_VIEW,
        CUSTOMISE,
        RTL,
        DELAYED_INIT
    }
}
